internal/k8s/controllers/http_route_controller.go

```diff
@@ -53,5 +61,123 @@ func (r *HTTPRouteReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
+// BackendRefs selectable by the To fields, but currently we just revalidate
+// all Routes allowed in the From Namespaces
+func (r *HTTPRouteReconciler) referencePolicyToRouteRequests(object client.Object) []reconcile.Request {
+	refPolicy := object.(*gateway.ReferencePolicy)
+
+	routes := r.getRoutesAffectedByReferencePolicy(refPolicy)
+	requests := []reconcile.Request{}
+
+	for _, route := range routes {
+		requests = append(requests, reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      route.Name,
+				Namespace: route.Namespace,
+			},
+		})
+	}
+
+	return requests
+}
+
+func (r *HTTPRouteReconciler) getRoutesAffectedByReferencePolicy(refPolicy *gateway.ReferencePolicy) []gateway.HTTPRoute {
+	matches := []gateway.HTTPRoute{}
+
+	// toSelectors := []fields.Selector{}
+	// for _, to := range refPolicy.Spec.To {
+	// 	// When empty, the Kubernetes core API group is inferred.
+	// 	group := "core/v1"
+	// 	if to.Group != "" {
+	// 		group = string(to.Group)
+	// 	}
+
+	// 	toSelectors = append(toSelectors, groupKindToFieldSelector(schema.GroupKind{
+	// 		Group: group,
+	// 		Kind:  string(to.Kind),
+	// 	}))
+	// }
+
+	routes := r.getReferencePolicyObjectsFrom(refPolicy)
+
+	// TODO: match only routes with BackendRefs selectable by a
+	// ReferencePolicyTo instead of appending all routes. This seems expensive,
+	// so not sure if it would actually improve performance or not.
+	matches = append(matches, routes...)
+
+	// for _, route := range routes {
+	// 	routeMatched := false
+
+	// 	// TODO: should this use reflection to handle xRoute types? seems expensive
+	// 	for _, rule := range route.Spec.Rules {
+	// 		for range rule.BackendRefs {
+	// 			// Check if backendRef.BackendObjectReference is selectable by
+	// 			// the requirements in any refPolicy.Spec.To
+	// 			for _, selector := range toSelectors {
+	// 				if selector.Matches(fields.Set{
+	//                  "kind": backendRef.BackendObjectReference.Kind
+	//                  "metadata.name": backendRef.BackendObjectReference.Name
+	// 				}) {
+	// 					routeMatched = true
+	// 					matches = append(matches, route)
+
+	// 					// Exit toSelectors loop early if route has already been matched
+	// 					if routeMatched {
+	// 						break
+	// 					}
+	// 				}
+	// 			}
+
+	// 			// Exit BackendRefs loop early if route has already been matched
+	// 			if routeMatched {
+	// 				break
+	// 			}
+	// 		}
+
+	// 		// Exit Rules loop early if route has already been matched
+	// 		if routeMatched {
+	// 			break
+	// 		}
+	// 	}
+	// }
+
+	return matches
+}
+
+func (r *HTTPRouteReconciler) getReferencePolicyObjectsFrom(refPolicy *gateway.ReferencePolicy) []gateway.HTTPRoute {
+	matches := []gateway.HTTPRoute{}
+
+	for _, from := range refPolicy.Spec.From {
+		// TODO: search by from.Group and from.Kind instead of assuming HTTPRoute
+		routes, err := r.Client.GetHTTPRoutesInNamespace(r.Context, string(from.Namespace))
```

Comment:
After a few hours of digging through the dynamic and unstructured packages, I don't think it's strictly possible to search for multiple kinds in a single API call with the go-client (would love to be proven wrong here though). The most generic thing I can figure out how to do is potentially use a mapper to search for the routes that are available on the cluster and list each of those found group version kinds and apply our logic from there, but I'm not sure that even makes sense at the moment when we're only supporting two kinds.

Reply:
I'm thinking it's not worth more digging vs. just having `r.Client.GetTCPRoutesInNamespace( ... )`. Could wait and dig more if we wind up hitting rule of three

Reply:
👍 to revisiting this when we get to implementing [TLSRoute](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.TLSRoute) or [UDPRoute](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.UDPRoute) - something more generic _might_ be possible with `unstructured.Unstructured` struct and `scheme.Runtime` using reflecting to convert from `kind` strings to types as described in https://iximiuz.com/en/posts/kubernetes-api-go-types-and-common-machinery/ (and likely adding a RouteController interface), but definitely not straightforward.

Implemented as `GetTCPRoutesInNamespace` in #162

Reply:
Rather than doing things with strings and `unstructured`, if you really want you could just make a generic `List`/`ListInNamespace` function on the interface and pass in the object you want retrieved directly. The k8s client uses reflection for retrieving lists/objects anyway. It'd handle something like `client.ListInNamespace(ctx, &gateway.HTTPRouteList{}, namespace)`. Which is a lot closer to the vanilla k8s client `List` method. `TCPRoute` would just then be `client.ListInNamespace(ctx, &gateway.TCPRouteList{}, namespace)`. The function signature would just specify a [`runtimeclient.ObjectList`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client#ObjectList).

The main two reasons you'll still want to wrap the k8s client call, even if it's a really thin wrapper over `List` are for:

1. `K8sError` wrapping for our retry middleware
2. Easy mocking for error injection for tests

I don't necessarily have much of a preference but think the per-type signatures are kind of nice in that you can just having the initialized structures passed back to you. Per-type also follow more of the pattern throughout most of the `gatewayclient` interface. But either way is fine with me. If you _do_ change it though, I'd do that in a follow-up PR at this point, no need to re-block.